sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/sns/PublishResultCoders.java

```diff
@@ -0,0 +1,129 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.aws.sns;
+
+import com.amazonaws.ResponseMetadata;
+import com.amazonaws.http.SdkHttpMetadata;
+import com.amazonaws.services.sns.model.PublishResult;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Optional;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.NullableCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.aws.coders.AwsCoders;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+/** Coders for SNS {@link PublishResult}. */
+public final class PublishResultCoders {
+
+  private static final Coder<String> MESSAGE_ID_CODER = StringUtf8Coder.of();
+  private static final Coder<ResponseMetadata> RESPONSE_METADATA_CODER =
+      AwsCoders.responseMetadata();
+
+  private PublishResultCoders() {}
+
+  /**
+   * Returns a new PublishResult coder which by default serializes only the messageId.
+   *
+   * @return the PublishResult coder
+   */
+  public static Coder<PublishResult> defaultPublishResult() {
+    return new PublishResultCoder(null, null);
+  }
+
+  /**
+   * Returns a new PublishResult coder which serializes the sdkResponseMetadata and sdkHttpMetadata,
+   * but does not include the HTTP response headers.
+   *
+   * @return the PublishResult coder
+   */
+  public static Coder<PublishResult> fullPublishResultNoHeaders() {
+    return new PublishResultCoder(RESPONSE_METADATA_CODER, AwsCoders.sdkHttpMetadata());
+  }
+
+  /**
+   * Returns a new PublishResult coder which serializes the sdkResponseMetadata and sdkHttpMetadata,
+   * including the HTTP response headers.
+   *
+   * @return the PublishResult coder
+   */
+  public static Coder<PublishResult> fullPublishResultIncludingHeaders() {
+    return new PublishResultCoder(RESPONSE_METADATA_CODER, AwsCoders.sdkHttpMetadata(true));
+  }
+
+  private static class PublishResultCoder extends Coder<PublishResult> implements Serializable {
+
+    private final Coder<ResponseMetadata> responseMetadataEncoder;
+    private final Coder<SdkHttpMetadata> sdkHttpMetadataCoder;
+
+    protected PublishResultCoder(
+        Coder<ResponseMetadata> responseMetadataEncoder,
+        Coder<SdkHttpMetadata> sdkHttpMetadataCoder) {
+      this.responseMetadataEncoder =
+          Optional.ofNullable(responseMetadataEncoder).map(NullableCoder::of).orElse(null);
```

Comment:
Why add this complexity here, just pass in a NullableCoder.of(RESPONSE_METADATA_CODER) and perform assignment here.